man/RNAsubopt.1.in
\" -*-nroff-*-
.\" .ER
.TH "RNASUBOPT" "l" "" "Ivo Hofacker" "Vienna RNA"
.SH "NAME"
RNAsubopt \- calculate suboptimal secondary structures of RNAs
.SH "SYNOPSIS"
\fBRNAsubopt [\-e\ \fIrange\fP] [\-ep \fIprange\fP] [\-s] [\-p\ \fIn\fP] [\-T\ \fItemp\fP] [\-d[0|1|2|3]] [\-4] [\-noGU] [\-noCloseGU] [\-P\ paramfile] [\-logML] [\-nsp\ pairs] [\-circ]

.SH "DESCRIPTION"
.I RNAsubopt
reads RNA sequences from stdin and (in the default \-e mode)
calculates all suboptimal secondary structures within a user defined
energy range above the minimum free energy (mfe).It prints the
suboptimal structures in bracket notation followed by the energy in
kcal/mol to stdout. Be careful, the number of structures returned
grows exponentially with both sequence length and energy range.

Alternatively, when used with the \-p option, RNAsubopt produces
Boltzmann weighted samples of secondary structures.

Sequences are read in the usual format, i.e. each sequence occupies a
single line, possibly preceded by a fasta-style header line of the form
.br
> name

.SH "OPTIONS"
.TP
.B \-e\ \fIrange\fP
Calculate suboptimal structures within \fIrange\fP kcal/mol of the mfe.
Default is 1.
.TP
.B \-s
Sort the structures by energy. Since the sort in is done in memory,
this becomes impractical when the number of structures produced goes
into millions. In such cases better pipe the output through `sort +1n'.
.TP
.B \-p \fIn\fP
Instead of producing all suboptimals in an energy range, produce a
random sample of \fIn\fP suboptimal structures, drawn with
probabilities equal to their Boltzmann weights via stochastic
backtracking in the partition function. The \-e and \-p options a
mutually exclusive.
.TP
.B \-d[0|1|2|3]
Change treatment of dangling ends, as in RNAfold and RNAeval.  The
default is \-d2 (as in partition function folding). If \-d1 or \-d3 are
specified the structures are generated as with \-d2 but energies are
re\-evaluated before printing.
.TP
.B \-logML
re\-calculate energies of structures using a logarithmic energy function for
multi\-loops before output. This option does not effect structure
generation, only the energies that is printed out. Since logML lowers
energies somewhat, some structures may be missing.
.TP
.B \-ep \fIprange\fP
Only print structures with energy within \fIprange\fR of the mfe. Useful in
conjunction with \-logML, \-d1 or \-d3: while the \-e option specifies the
range before energies are re\-evaluated, \-ep specifies the maximum energy
after re\-evaluation.
.TP
.B \-noLP
Only produce structures without lonely pairs (helices of length 1). This
reduces the number of structures drastically and should therefore be used
for longer sequences and larger energy ranges.
.TP
.B \-circ
Assume a circular (instead of linear) RNA molecule.
.PP
The \fB\-T\fP, \fB\-4\fP, \fB\-noGU\fP, \fB\-noCloseGU\fP,
\fB\-P\fP, \fB\-nsp\fP, options work as in RNAfold.


.SH "REFERENCES"
Please cite:
.br
S. Wuchty, W. Fontana, I. L. Hofacker and P. Schuster
"Complete Suboptimal Folding of RNA and the Stability of
Secondary Structures",
Biopolymers, 49, 145\-165 (1999)

.SH "VERSION"
This man page documents version @VERSION@ Vienna RNA Package.

.SH "AUTHORS"
Ivo L Hofacker, Stefan Wuchty, Walter Fontana.
.br
Send comments and bug reports to <rna@tbi.univie.ac.at>
\"  LocalWords:  RNASUBOPT RNAsubopt suboptimal RNAs fBRNAsubopt fIrange fP ep
\"  LocalWords:  prange lodos fItemp noGU noCloseGU paramfile logML nsp stdin
\"  LocalWords:  mfe kcal mol stdout TP RNAfold RNAeval multi fIprange noLP fB
\"  LocalWords:  br Wuchty Fontana Hofacker Schuster Biopolymers Ivo


NEWS
See Changelog for details.

Upcoming 
-----------
- Updated parameters for both RNA and DNA using the data compiled by 
  David Mathews <David_Mathews@urmc.rochester.edu>. 

Version 1.6
------------
- The RNAforester program for tree-alignments of RNA structures is now
  distributed with the Vienna RNA package, see the RNAforester
  subdirectory for more information. RNAforester was written by Matthias
  Hoechsmann <mhoechsm@techfak.uni-bielefeld.de> 
- The Kinfold program for stochastic simulation of folding trajectories is
  now included in the package, see the Kinfold subdirectory.
- cofolding of two structures now supports suboptimal folding and 
  partition function folding (still experimental). 
  ATTENTION: Energies of hybrid structures now inlcude the Duplex-initiation
  energy, which was neglected in previous version.
- RNAplfold is a partition function variant of RNALfold. It computes the
  mean probability of a (local) base pair averaged over all sequence
  windows that contain the pair.
- new utilities to color alignments and consensus structures  
- RNAfold -p now computes the centroid structure
- CAUTION: ensemble diversities in version <1.6.5 are off by a factor 2

Version 1.5pre 
-----------
- ViennaRNA now uses autoconfig generated configure scripts for even better
  portability (should compile on any UNIX, Linux, MacOS X, Windows with
  Cygwin).
- The new RNAalifold program predicts consensus structures for a set of
  aligned sequences.
- Complete suboptimal folding is now integrated in the library.
- Beginning support for co-folding of two strands: energy_of_struct() and
  RNAeval can now compute energies of duplex structures.
- RNAcofold predicts hybrid structures of two RNA strands
- RNAduplex predicts hybrid structures, while allowing only inter-molecular
  base pairs (useful for finding potential binding sites)
- RNALfold predicts locally stable structures in long sequences.
- Major changes to Perl module. See the pod documentation (perldoc RNA).
- RNAsubopt can do stochastic backtracking to produce samples of suboptimal
  structures with Boltzmann statistics.
- New utilities to rotate secondary structure plots and annotate them with
  reliability data.
- Various small bug fixes

Version 1.4
-----------
- New Turner parameters as described in Mathews et.al. JMB v288, 1999.
  Small changes to format of parameter files (old param files won't work!)
- mfe and suboptimal folding will produce only structures without isolated
  pairs if noLonelyPairs=1 (-noLP option), for partition function folding
  pairs that can only occur as isolated pairs are not formed.
- setting dangles=3 (-d3 option) will allow co-axial stacking of adjacent
  helices in mfe folding and energy_of_struct().

Version 1.3.1
-------------
- RNAheat would produce spikes in the specific heat because dangling end
  energies did not go smoothly to 0. 
- PS dot plots now have an option to use a log scale (edit _dp.ps file
  and set logscale to true).

Version 1.3
-----------
- Secondary structure plots now use E. Bruccoleri's naview routines for
  layout by default. New utility RNAplot produces secondary structure plots 
  from structures in bracket notation with several options.
- New -d2 option in RNAfold and RNAeval sets dangles=2, which makes 
  energy_of_struct() and fold() treat dangling ends as in pf_fold().
  -noLP option in RNAfold etc sets noLonelyPairs=1, which avoids most
  structures containing lonely base pairs (helices of length 1).
- new utility functions pack_structure() unpack_structure() make_pair_table()
  and bp_distance().  RNAdistance adds bp_distance() via -DP switch.
- First release of RNAsubopt for complete suboptimal folding.
- fixed bug in asymmetry penalty for interior loops. 
- Default compilation now uses doubles for partition function folding.

Version 1.2.1
-------------
- Fixed bug in version 1.2 of the RNAheat program causing overflow errors
  for most input sequences. 
- The PS_dot_plot() and PS_rna_plot() routines now return an int. The return 
  value is 0 if the file could not be written, 1 otherwise.
- This version contains the alpha version of a perl5 module, which let's you
  access all the capabilities of the Vienna RNA library from perl scripts.

Version 1.2
-----------
- New energy parameters from (Walter et.al 1994). 
- Energy parameters can be read from file.
- RNAeval and energy_of_struct() support logarithmic energy function for 
  multi-loops.
- gmlRNA() produces secondary structure drawing in gml (Graph Meta Language). 
- Many bug fixes.


Progs/RNAsubopt.c
/* Last changed Time-stamp: <2007-12-05 13:56:02 ronny> */
/*
		Ineractive access to suboptimal folding

			   c Ivo L Hofacker
			  Vienna RNA package
*/
#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <ctype.h>
#include <string.h>
#include "part_func.h"
#include "fold.h"
#include "fold_vars.h"
#include "utils.h"
#include "subopt.h"
extern void  read_parameter_file(const char fname[]);
extern int   st_back;
/*@unused@*/
static char UNUSED rcsid[] = "$Id: RNAsubopt.c,v 1.14 2007/12/05 13:04:08 ivo Exp $";

#define PRIVATE static

static char  scale[] = "....,....1....,....2....,....3....,....4"
		       "....,....5....,....6....,....7....,....8";

extern double print_energy;
PRIVATE void usage(void);
/*--------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
   char  *line;
   char *sequence;
   char *structure = NULL;
   char  fname[21];
   char  *ParamFile = NULL;
   char  *ns_bases = NULL, *c;
   int   i, length, l, sym, r;
   int   istty;
   double deltaf, deltap=0;
   int delta=100;
   int n_back = 0;
   int noconv = 0;
   int circ=0;

   do_backtrack = 1;
   dangles = 2;
   for (i=1; i<argc; i++) {
      if (argv[i][0]=='-')
	switch ( argv[i][1] )
	  {
	  case 'T':  if (argv[i][2]!='\0') usage();
	    if(i==argc-1) usage();
	    r=sscanf(argv[++i], "%lf", &temperature);
	    if (r!=1) usage();
	    break;
	  case 'p':
	    if (argv[i][2]!='\0') usage();
	    if(i==argc-1) usage();
	    (void) sscanf(argv[++i], "%d", &n_back);
	    init_rand();
	    break;
	  case 'n':
	    if ( strcmp(argv[i], "-noGU" )==0) noGU=1;
	    if ( strcmp(argv[i], "-noCloseGU" ) ==0) no_closingGU=1;
	    if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	    if ( strcmp(argv[i], "-nsp") ==0) {
	      if (i==argc-1) usage();
	      ns_bases = argv[++i];
	    }
	    if ( strcmp(argv[i], "-noconv")==0) noconv=1;
	    break;
	  case '4':
	    tetra_loop=0;
	    break;
	  case 'C':
	    fold_constrained=1;
	    break;
	  case 'd': dangles=0;
	    if (argv[i][2]!='\0') {
	      r=sscanf(argv[i]+2, "%d", &dangles);
	      if (r!=1) usage();
	    }
	    break;
	  case 'P':
	    if (i==argc-1) usage();
	    ParamFile = argv[++i];
	    break;
	  case 's':
	    subopt_sorted=1;
	    break;
	  case 'l':
	    if (strcmp(argv[i],"-logML")==0) {
	      logML=1;
	      break;
	    }
	    else usage();
	    break;
	  case 'e':
	    if (i>=argc-1) usage();
	    if (strcmp(argv[i],"-ep")==0)
	      r=sscanf(argv[++i], "%lf", &deltap);
	    else {
	      r=sscanf(argv[++i], "%lf", &deltaf);
	      delta = (int) (0.1+deltaf*100);
	    }
	    if (r!=1) usage();
	    break;
		case 'c':
			if ( strcmp(argv[i], "-circ")==0) circ=1;
			break;
	  default: usage();
	  }
   }

   if (ParamFile != NULL)
     read_parameter_file(ParamFile);

   if (ns_bases != NULL) {
      nonstandards = space(33);
      c=ns_bases;
      i=sym=0;
      if (*c=='-') {
	 sym=1; c++;
      }
      while (*c) {
	if (*c!=',') {
	  nonstandards[i++]=*c++;
	  nonstandards[i++]=*c;
	  if ((sym)&&(*c!=*(c-1))) {
	    nonstandards[i++]=*c;
	    nonstandards[i++]=*(c-1);
	  }
	}
	c++;
      }
   }
   istty = isatty(fileno(stdout))&&isatty(fileno(stdin));
   if ((fold_constrained)&&(istty)) {
      printf("Input constraints using the following notation:\n");
      /* printf("| : paired with another base\n"); */
      printf(". : no constraint at all\n");
      printf("x : base must not pair\n");
   }

   do {				/* main loop: continue until end of file */
      if (istty) {
	 printf("\nInput string (upper or lower case); @ to quit\n");
	 printf("%s\n", scale);
      }
      fname[0]='\0';
      if ((line = get_line(stdin))==NULL) break;

      /* skip comment lines and get filenames */
      while ((*line=='*')||(*line=='\0')||(*line=='>')) {
	if (*line=='>')
	  (void) sscanf(line, ">%20s", fname);
	free(line);
	if ((line = get_line(stdin))==NULL) break;;
      }

      if ((line==NULL) || strcmp(line, "@") == 0) break;

      sequence = (char *) space(strlen(line)+1);
      (void) sscanf(line,"%s",sequence);
      free(line);
      length = (int) strlen(sequence);

      structure = (char *) space((unsigned) length+1);
      if (fold_constrained) {
	char *cstruc;
	cstruc = get_line(stdin);
	if (cstruc!=NULL) {
	  strncpy(structure, cstruc, length);
	  for (i=0; i<length; i++)
	    if (structure[i]=='|')
	      nrerror("constraints of type '|' not allowed");
	  free(cstruc);
	}
      }

      for (l = 0; l < length; l++) {
	sequence[l] = toupper(sequence[l]);
	if (!noconv && sequence[l] == 'T') sequence[l] = 'U';
      }
      if (istty)
	printf("length = %d\n", length);

      if (logML!=0 || dangles==1 || dangles==3)
	if (deltap<=0) deltap=delta/100. +0.001;
      if (deltap>0)
	print_energy = deltap;

      /* first lines of output (suitable  for sort +1n) */
      if (fname[0] != '\0')
	printf("> %s [%d]\n", fname, delta);

      if (n_back>0) {
	int i;
	double mfe, kT;
	char *ss;
	st_back=1;
	ss = (char *) space(strlen(sequence)+1);
	strncpy(ss, structure, length);
	mfe = (circ) ? circfold(sequence, ss) : fold(sequence, ss);
	kT = (temperature+273.15)*1.98717/1000.; /* in Kcal */
	pf_scale = exp(-(1.03*mfe)/kT/length);
	strncpy(ss, structure, length);
    /* we are not interested in the free energy but in the bppm, so we drop */
    /* free energy into the void */
    (circ) ? (void) pf_circ_fold(sequence, ss)	: (void) pf_fold(sequence, ss);
	free(ss);
	for (i=0; i<n_back; i++) {
	  char *s;
	  s = (circ) ? pbacktrack_circ(sequence) : pbacktrack(sequence);
	  printf("%s\n", s);
	  free(s);
	}
	free_pf_arrays();
      } else {
	(circ) ? subopt_circ(sequence, structure, delta, stdout) : subopt(sequence, structure, delta, stdout);
      }
      (void)fflush(stdout);

      free(sequence);
      free(structure);
   } while (1);
   return 0;
}

PRIVATE void usage(void)
{
   nrerror("usage: "
	   "RNAsubopt [-e range] [-ep prange] [-s] [-p num] [-logML]\n"
	   "          [-C] [-T temp] [-4] [-d[2]] [-noGU] [-noCloseGU]\n"
	   "          [-noLP] [-P paramfile] [-nsp pairs] [-circ]");
}
